Brand address create/edit actions migration

zuk3975 opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrate Catalog > Brands & Suppliers > Address create/edit actions
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #10544
| How to test?  | See below

## How to test

Edit and create actions should work as in legacy page except following changes:
- :warning: On create/edit action enabled selecting and reseting brand
- In form field 'others' fixed illegal characters hint  according to its validation rules.

Testing forms can be accessed from listing page at `/admin-dev/index.php/sell/catalog/brands/`, follow the click on Add or Edit button.
Or directly via links
- `/admin-dev/index.php/sell/catalog/brands/addresses/new`
- `/admin-dev/index.php/sell/catalog/brands/addresses/{addressId}/edit`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12795)
<!-- Reviewable:end -->

zuk3975 committed:
Initial commit. Routes and empty create/edit methods

zuk3975 committed:
Render empty address create form

zuk3975 committed:
Added configurable form choices provider

zuk3975 committed:
Enabled javascript for handling dynamic country state option

zuk3975 committed:
Some services renaming

zuk3975 committed:
Implemented command handler for address creation

zuk3975 committed:
Minor code optimization

zuk3975 committed:
Implemented address edit handler

zuk3975 committed:
Added ManufacturerId VO, fixed illegal chars hint in twig

zuk3975 committed:
EditAddress and formData handlers

zuk3975 committed:
Added Exception message in controller

sarjon committed:
fix translations for form choices

sarjon committed:
extract states toggler into reusable JS component

sarjon committed:
disable manufacturer selection when editing address

sarjon committed:
redirect user to manufacturers list after editing address

sarjon committed:
code cleanup

sarjon committed:
minor improvements

sarjon committed:
fix naming conventions

zuk3975 committed:
[WIP] fixing edit/create form manufacturer choice

zuk3975 commented on src/Adapter/Address/CommandHandler/AddManufacturerAddressHandler.php:
```diff
@@ -0,0 +1,86 @@
+    {
+        $address = new Address();
+        if (null !== $command->getManufacturerId()) {
+            $address->id_manufacturer = $command->getManufacturerId()->getValue();
```
@matks 
Manufacturer null value should be accepted as a valid choice (as colinegin mentioned in #9337).
When choice is not changed in edit form, this value will be null, as well as if choice where changed to "no manufacturer" (null too).
I.e. Address has manufacturer and user wants to remove it - user selects "no manufacturers". $manufacturerId will get null value and this handler would not update that field.
What would be the best way handling this?

zuk3975 committed:
Fixed typo mistake in form constraint

zuk3975 commented on src/Adapter/Address/CommandHandler/EditManufacturerAddressHandler.php:
```diff
@@ -0,0 +1,110 @@
+    private function populateAddressWithData(Address $address, $command)
+    {
+        if (null !== $command->getManufacturerId()) {
+            $address->id_manufacturer = $command->getManufacturerId();
```
@matks,
`EditManufacturerAddressCommand` accepts parameter `manufacturerId` which can be `int` in case you want to update it or `null` if you want to reset it.

The problem is when we have `Address` with lets say Manufacturer id `2` and we only want to update Addresse's city (or any other field except manufacturer). So, we create command like so:

```php
$command = new EditManufacturerAddressCommand(2);
$command->setCity('New city');

// when dispatching $command we don't expect manufacturer id to be changed
// as we did not set it explicitly.
$bus->handle($command);
```

However, in this command Manufacturer id is `null` as we didn't set it. But when handling command, we never know if `null` was set intentionally (to reset addresse's manufacturer) or left untouched as both values ends up being `null`. Thus, we don't know whether Addresse's manufacturer Id should be reset or kept as is.

Do you have any ideas of how we could tell handler if manufacturer's id for address has changed or not?

matks replied:
First question: so this means we can reset an address Manufacturer Id ? I'd say yes, since you want to handle this usecase.

I think we can acknowledge the intentional passing of `null` by keeping a flag:
```
private $manufacturerIdShouldBeReset = false;

public function setManufacturerId($manufacturerId)
    {
        if (null !== $manufacturerId) {
            $manufacturerId = new ManufacturerId($manufacturerId);
            $this->manufacturerId = $manufacturerId;
        } else {
            $this->manufacturerIdShouldBeReset = true;
        }
    }
```

Then in the handler, if `$command->shouldManufacturerIdBeReset` is true you reset it, else you ignore it.

And in order to increase DX, we can even a dedicated function (so people have 2 ways to reset the manufacturer ID: use regular setter with `null` or use this dedicated function)
```
public function setManufacturerIdToBeReset()
    {
        $this->manufacturerIdShouldBeReset = true;
    }
```

zuk3975 committed:
Made manufacturer choice optional in address form

zuk3975 committed:
Added missing form rest

matks commented on admin-dev/themes/new-theme/js/components/country-state-selection-toggler.js:
```diff
@@ -0,0 +1,85 @@
+ *   ...
+ * </select>
+ *
+ * <!-- If selected country does not have states, then this block will be hidden -->
```
```suggestion
 * <!-- If selected country does not have state, then this block will be hidden -->
```

zuk3975 replied:
@matks i think in this case should be states.
It supposed to check whether the country has states at all, not for checking specific state.

matks commented on admin-dev/themes/new-theme/js/components/country-state-selection-toggler.js:
```diff
@@ -0,0 +1,85 @@
+      data: {
+        id_country: this.$countryInput.val(),
+      }
+    }).then((response) => {
```
Can you add a `.fail()` usecase to display a nice error message to user if the ajax call fails (HTTP 500, HTTP 400) or timeouts ?

jolelievre replied:
agreed, could you add the `.fail()` using `showErrorMessage` in it please?

jolelievre replied:
@zuk3975 still this fail management to do

zuk3975 replied:
@jolelievre I use .catch() it does the same thing in this case.

jolelievre replied:
I didn't know catch as a replacement for fail, thanks.
However I believe it is not managed by our version of jquery, it was introduced in v3 https://api.jquery.com/deferred.catch/ and seems like we are using jQuery v1.11 ....

matks commented on src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php:
```diff
@@ -0,0 +1,328 @@
+     */
+    public function setManufacturerId($manufacturerId)
+    {
+        if (!is_int($manufacturerId) || 0 > $manufacturerId) {
```
Should not we use again a `assertIsNullOrNonNegativeInt()` method ? An Edit Address Command cannot reset the manufacturer ID ?

matks commented on src/Core/Domain/Manufacturer/ValueObject/ManufacturerId.php:
```diff
@@ -0,0 +1,75 @@
+/**
+ * Provides manufacturer id
+ */
+class ManufacturerId
```
This is going to create conflicts 😭

matks commented on src/Core/Form/ConfigurableFormChoiceProviderInterface.php:
```diff
@@ -0,0 +1,40 @@
+     *
+     * @return array
+     */
+    public function getChoices(array $options);
```
Mmmm 🤔 I see why you introduced this. So this is a opponent of `FormChoiceProviderInterface`.
Now that I see it I wonder why we did not introduce it sooner.

@mickaelandrieu @sarjon @tomas862 wdyt ?

matks replied:
I'm worried there is somewhere some generic code, like a Factory, expecting a `FormChoiceProviderInterface`.

sarjon replied:
What do you mean @matks?

For choices that don't need any options, we keep using `FormChoiceProviderInterface`, however for choices that can be configured, we can use `ConfigurableFormChoiceProviderInterface` instead.

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php:
```diff
@@ -0,0 +1,64 @@
+class StateController extends FrameworkBundleAdminController
+{
+    /**
+     * Provides country states in json response
```
No `AdminSecurity` but this expected here. Can you confirm that it still requires to be logged in (so a visitor or an anonymous cannot access this route) ?

jolelievre replied:
Did you check if it is accessible offline?

zuk3975 replied:
@jolelievre  When trying to reach it without being logged in it redirects to login page

jolelievre replied:
Great!

matks reviewed (changes requested):
Nice work 👏 some minor issues

zuk3975 committed:
Some code optimization and fixes

rokaszygmantas commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php:
```diff
@@ -0,0 +1,166 @@
+     */
+    public function editAddressAction(Request $request, $addressId)
+    {
+        $addressId = (int) $addressId;
```
either this line or further int casting of `$addressId` is redundant in this action

rokaszygmantas commented on src/Core/Form/IdentifiableObject/DataProvider/ManufacturerAddressFormDataProvider.php:
```diff
@@ -0,0 +1,93 @@
+     */
+    private $defaultCountryId;
+
+    public function __construct(
```
missing phpdoc

zuk3975 committed:
Some code optimization

zuk3975 committed:
Removed unnecessary security annotation

zuk3975 committed:
rebase fixes

zuk3975 committed:
rebase fixes

marionf commented:
@zuk3975 

1. For the "other" field, I have the message "This value is not valid"

![capture d'écran_1242](https://user-images.githubusercontent.com/13449658/55092342-bbff9200-50b2-11e9-9dca-1ad4dc272c2a.png)


2. When I add a new brand address, I can choose "--", which wasn't possible in legacy

![capture d'écran_1243](https://user-images.githubusercontent.com/13449658/55092541-05e87800-50b3-11e9-9f41-dc3e159e6b51.png)

3. When I edit a brand address, I can change the brand,  which wasn't possible in legacy

zuk3975 committed:
Added TypedRegexConstraints

zuk3975 commented:
> @zuk3975
> 
> 1. For the "other" field, I have the message "This value is not valid"
> 
> ![capture d'écran_1242](https://user-images.githubusercontent.com/13449658/55092342-bbff9200-50b2-11e9-9dca-1ad4dc272c2a.png)
> 
> 2. When I add a new brand address, I can choose "--", which wasn't possible in legacy
> 
> ![capture d'écran_1243](https://user-images.githubusercontent.com/13449658/55092541-05e87800-50b3-11e9-9f41-dc3e159e6b51.png)
> 
> 3. When I edit a brand address, I can change the brand,  which wasn't possible in legacy

@marionf, 1 - I fixed the `Other` field also fixed wrong hints according to validation.
2 & 3. These changes are intentional. We have discussed it here `https://github.com/PrestaShop/PrestaShop/pull/9337/files/f56b478847b138150a9785c741e438cd07c6339d#r262988757`

marionf commented:
All right, thank you @zuk3975 :+1:

zuk3975 committed:
Fixed manufacturers/addresses js entry points

zuk3975 committed:
Modified webpack entry points for manufacturers/addresses

jolelievre reviewed (commented):
@zuk3975 only two remaining feedback, a little rebase and we're good

zuk3975 commented:
@matks rebased once again

zuk3975 committed:
Added fieldLangs validation in add/edit handlers

zuk3975 committed:
Fixed validateFieldsLang methods typos

zuk3975 commented:
@matks rebased once again :D

marionf commented:
@zuk3975 

State dropdown is empty, no matter the country selected

![capture d'écran_1263](https://user-images.githubusercontent.com/13449658/55233960-5504e900-522a-11e9-8824-44062fecb56c.png)

zuk3975 commented:
> @zuk3975
> 
> State dropdown is empty, no matter the country selected
> 
> ![capture d'écran_1263](https://user-images.githubusercontent.com/13449658/55233960-5504e900-522a-11e9-8824-44062fecb56c.png)

@marionf It seems to be working locally once again. I Rebuilt and pushed assets and hope it was the cause :)

marionf commented:
@zuk3975 State is still empty for me :(

zuk3975 commented:
> @zuk3975 State is still empty for me :(

@marionf,  Can you provide browser console screenshot when you test this? Because its issue with js and might see some errors there.

marionf commented:
@zuk3975 

There is nothing in the console

![capture d'écran_1265](https://user-images.githubusercontent.com/13449658/55242275-2d6b4c00-523d-11e9-9e57-91847ed88f3a.png)

matks committed:
Handle dynamic states loading for Add/Edit manufacturer address selection

matks commented:
@zuk3975 I did not see the javascript handling of country switch for states, so marionf's report seems relevant 🤔

Something like this does the job: https://github.com/matks/PrestaShop/commit/f339135ec8d0684fe281ee391c25bbeabd0f421c

However then there is validation issue, because the choices that are loaded [here](https://github.com/PrestaShop/PrestaShop/pull/12795/files#diff-6f21bb46d5d2e6916d3a7e4c5938b447R224) are the ones linked to the context country. So if you're in France, the ChoiceType loads French states choices (which is empty) even if you select another country. We need to load the choices related to the selected country (in the form) rather than the context country.

matks committed:
Handle dynamic state validation for Add/Edit manufacturer address

matks committed:
Build assets and add missing default value to toggler function

matks commented:
@marionf State drop-down issue is solved \o/

zuk3975 committed:
Rebuilt assets

matks commented:
Thank you @zuk3975
